File: shared_swift/AudioRenderer.swift
/*
 * Copyright 2025 LiveKit
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import AVFoundation
import WebRTC

#if os(macOS)
import Cocoa
import FlutterMacOS
#else
import Flutter
import UIKit
#endif

public class AudioRenderer: NSObject {
    public let rendererId: String
    public let format: AVAudioFormat // Target format

    private var eventSink: FlutterEventSink?
    private var channel: FlutterEventChannel?

    private var converter: AudioConverter?

    // Weak ref
    private weak var _track: AudioTrack?

    public init(track: AudioTrack,
                binaryMessenger: FlutterBinaryMessenger,
                rendererId: String,
                format: AVAudioFormat)
    {
        _track = track
        self.rendererId = rendererId
        self.format = format

        super.init()
        _track?.add(audioRenderer: self)

        let channelName = "io.livekit.audio.renderer/channel-" + rendererId
        channel = FlutterEventChannel(name: channelName, binaryMessenger: binaryMessenger)
        channel?.setStreamHandler(self)
    }

    func detach() {
        _track?.remove(audioRenderer: self)
    }

    deinit {
        detach()
    }
}

extension AudioRenderer: FlutterStreamHandler {
    public func onListen(withArguments _: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    public func onCancel(withArguments _: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}

public extension AVAudioPCMBuffer {
    /// Serializes audio data as raw interleaved bytes.
    ///
    /// Mono buffers are copied directly.
    /// Multi-channel buffers are interleaved in sample order.
    ///
    /// Uses `FlutterStandardTypedData` to send a binary payload.
    func serializeAsBytes() -> [String: Any] {
        let channels = Int(format.channelCount)
        let frames = Int(frameLength)

        var result: [String: Any] = [
            "sampleRate": UInt(format.sampleRate),
            "channels": UInt(channels),
            "frameLength": UInt(frames),
        ]

        if let int16ChannelData {
            let data: Data
            if channels == 1 {
                // Fast path for mono.
                data = Data(bytes: int16ChannelData[0], count: frames * MemoryLayout<Int16>.size)
            } else {
                // Interleave channels
                var bytes = Data(count: frames * channels * MemoryLayout<Int16>.size)
                bytes.withUnsafeMutableBytes { raw in
                    let out = raw.bindMemory(to: Int16.self)
                    for frame in 0..<frames {
                        for ch in 0..<channels {
                            out[frame * channels + ch] = int16ChannelData[ch][frame]
                        }
                    }
                }
                data = bytes
            }
            result["data"] = FlutterStandardTypedData(bytes: data)
            result["commonFormat"] = "int16"
        } else if let floatChannelData {
            let data: Data
            if channels == 1 {
                data = Data(bytes: floatChannelData[0], count: frames * MemoryLayout<Float32>.size)
            } else {
                var bytes = Data(count: frames * channels * MemoryLayout<Float32>.size)
                bytes.withUnsafeMutableBytes { raw in
                    let out = raw.bindMemory(to: Float32.self)
                    for frame in 0..<frames {
                        for ch in 0..<channels {
                            out[frame * channels + ch] = floatChannelData[ch][frame]
                        }
                    }
                }
                data = bytes
            }
            result["data"] = FlutterStandardTypedData(bytes: data)
            result["commonFormat"] = "float32"
        } else if let int32ChannelData {
            let data: Data
            if channels == 1 {
                data = Data(bytes: int32ChannelData[0], count: frames * MemoryLayout<Int32>.size)
            } else {
                var bytes = Data(count: frames * channels * MemoryLayout<Int32>.size)
                bytes.withUnsafeMutableBytes { raw in
                    let out = raw.bindMemory(to: Int32.self)
                    for frame in 0..<frames {
                        for ch in 0..<channels {
                            out[frame * channels + ch] = int32ChannelData[ch][frame]
                        }
                    }
                }
                data = bytes
            }
            result["data"] = FlutterStandardTypedData(bytes: data)
            result["commonFormat"] = "int32"
        } else {
            result["data"] = FlutterStandardTypedData(bytes: Data())
            result["commonFormat"] = "unknown"
        }

        return result
    }
}

extension AudioRenderer: RTCAudioRenderer {
    public func render(pcmBuffer: AVAudioPCMBuffer) {
        guard let eventSink else { return }

        // Create or update converter if needed
        if converter == nil || pcmBuffer.format != converter!.inputFormat || format != converter!.outputFormat {
            converter = AudioConverter(from: pcmBuffer.format, to: format)
        }

        let convertedBuffer = converter!.convert(from: pcmBuffer)

        guard convertedBuffer.frameLength == UInt32(format.sampleRate / 100) else {
            print("Converted buffer frame length does not match target format sample rate: \(convertedBuffer.frameLength) != \(format.sampleRate / 100) skipping this frame...")
            return
        }

        let serializedBuffer = convertedBuffer.serializeAsBytes()

        // Send the result to Flutter on the main thread
        DispatchQueue.main.async {
            eventSink(serializedBuffer)
        }
    }
}

File: android/src/main/kotlin/io/livekit/plugin/AudioRenderer.kt
/*
 * Copyright 2024 LiveKit, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.livekit.plugin

import android.os.Handler
import android.os.Looper
import android.util.Log
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import org.webrtc.AudioTrackSink
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * AudioRenderer for capturing audio data from WebRTC tracks and streaming to Flutter
 * Similar to iOS AudioRenderer implementation
 */
class AudioRenderer(
  private val audioTrack: LKAudioTrack,
  private val binaryMessenger: BinaryMessenger,
  private val rendererId: String,
  private val targetFormat: RendererAudioFormat
) : EventChannel.StreamHandler, AudioTrackSink {
  companion object {
    private const val TAG = "LKAudioRenderer"
  }

  private var eventChannel: EventChannel? = null
  private var eventSink: EventChannel.EventSink? = null
  private var isAttached = false
  private var droppedFrameCount = 0L

  private val handler: Handler by lazy {
    Handler(Looper.getMainLooper())
  }

  init {
    val channelName = "io.livekit.audio.renderer/channel-$rendererId"
    eventChannel = EventChannel(binaryMessenger, channelName)
    eventChannel?.setStreamHandler(this)

    // Attach to the audio track
    audioTrack.addSink(this)
    isAttached = true
  }

  fun detach() {
    if (isAttached) {
      audioTrack.removeSink(this)
      isAttached = false
    }
    eventChannel?.setStreamHandler(null)
    eventSink = null
  }

  override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
    eventSink = events
  }

  override fun onCancel(arguments: Any?) {
    eventSink = null
  }

  override fun onData(
    audioData: ByteBuffer,
    bitsPerSample: Int,
    sampleRate: Int,
    numberOfChannels: Int,
    numberOfFrames: Int,
    absoluteCaptureTimestampMs: Long
  ) {
    eventSink?.let { sink ->
      val convertedData = try {
        // Convert audio data to the target format
        convertAudioData(
          audioData,
          bitsPerSample,
          sampleRate,
          numberOfChannels,
          numberOfFrames
        )
      } catch (e: Exception) {
        logDroppedFrame("Audio conversion exception: ${e.message}")
        null
      }

      if (convertedData == null) {
        return@let
      }

      // Send to Flutter on the main thread
      handler.post {
        sink.success(convertedData)
      }
    }
  }

  /**
   * Converts audio data to raw interleaved bytes with resampling.
   *
   * Pipeline: read int16 → resample → channel reduce → format convert (int16/float32)
   */
  private fun convertAudioData(
    audioData: ByteBuffer,
    bitsPerSample: Int,
    sampleRate: Int,
    numberOfChannels: Int,
    numberOfFrames: Int
  ): Map<String, Any>? {
    // WebRTC AudioTrackSink always delivers 16-bit signed int16 PCM.
    if (bitsPerSample != 16) {
      logDroppedFrame("Unsupported bitsPerSample: $bitsPerSample (expected 16)")
      return null
    }
    if (numberOfChannels <= 0) {
      logDroppedFrame("Invalid numberOfChannels: $numberOfChannels")
      return null
    }
    if (numberOfFrames <= 0) {
      logDroppedFrame("Invalid numberOfFrames: $numberOfFrames")
      return null
    }

    val bytesPerFrame = numberOfChannels * 2
    val buffer = audioData.duplicate()
    buffer.order(ByteOrder.LITTLE_ENDIAN)
    buffer.rewind()

    val availableBytes = buffer.remaining()
    if (availableBytes <= 0) {
      logDroppedFrame("No audio payload bytes available")
      return null
    }

    val expectedBytes = numberOfFrames.toLong() * bytesPerFrame.toLong()
    val srcFrames = if (expectedBytes <= availableBytes.toLong()) {
      numberOfFrames
    } else {
      val availableFrames = availableBytes / bytesPerFrame
      if (availableFrames <= 0) {
        logDroppedFrame(
          "Insufficient bytes for one frame (available=$availableBytes, bytesPerFrame=$bytesPerFrame)"
        )
        return null
      }
      logDroppedFrame("Short audio payload; truncating frames from $numberOfFrames to $availableFrames")
      availableFrames
    }

    // Step 1: Read source int16 samples into ShortArray
    val src = ShortArray(srcFrames * numberOfChannels)
    for (i in src.indices) {
      src[i] = buffer.short
    }

    // Step 2: Resample to target sample rate
    val resampleResult = AudioResampler.resample(
      src, srcFrames, sampleRate, targetFormat.sampleRate, numberOfChannels
    )
    val resampled = resampleResult.samples
    val outFrames = resampleResult.frameCount

    if (outFrames <= 0) {
      logDroppedFrame("Resampled frame count is 0")
      return null
    }

    // Step 3: Channel reduction + format conversion
    val requestedChannels = targetFormat.numberOfChannels.coerceAtLeast(1)
    val outChannels = requestedChannels.coerceAtMost(numberOfChannels)

    val result = mutableMapOf<String, Any>(
      "sampleRate" to targetFormat.sampleRate,
      "channels" to outChannels,
      "frameLength" to outFrames,
    )

    when (targetFormat.commonFormat) {
      "float32" -> {
        result["commonFormat"] = "float32"
        val out = ByteArray(outFrames * outChannels * 4)
        val outBuf = ByteBuffer.wrap(out).order(ByteOrder.LITTLE_ENDIAN)
        for (f in 0 until outFrames) {
          for (ch in 0 until outChannels) {
            val sample = resampled[f * numberOfChannels + ch].toFloat() / Short.MAX_VALUE
            outBuf.putFloat((f * outChannels + ch) * 4, sample)
          }
        }
        result["data"] = out
      }
      else -> {
        result["commonFormat"] = "int16"
        if (outChannels == numberOfChannels) {
          // Fast path: no channel reduction — bulk copy resampled data
          val out = ByteArray(outFrames * outChannels * 2)
          val outBuf = ByteBuffer.wrap(out).order(ByteOrder.LITTLE_ENDIAN)
          for (i in 0 until outFrames * outChannels) {
            outBuf.putShort(i * 2, resampled[i])
          }
          result["data"] = out
        } else {
          // Channel reduction: keep first outChannels
          val out = ByteArray(outFrames * outChannels * 2)
          val outBuf = ByteBuffer.wrap(out).order(ByteOrder.LITTLE_ENDIAN)
          for (f in 0 until outFrames) {
            for (ch in 0 until outChannels) {
              outBuf.putShort(
                (f * outChannels + ch) * 2,
                resampled[f * numberOfChannels + ch]
              )
            }
          }
          result["data"] = out
        }
      }
    }

    return result
  }

  private fun logDroppedFrame(reason: String) {
    droppedFrameCount += 1
    if (droppedFrameCount <= 5 || droppedFrameCount % 100 == 0L) {
      Log.w(TAG, "Dropping audio frame #$droppedFrameCount for rendererId=$rendererId: $reason")
    }
  }
}

/**
 * Audio format specification for the renderer
 */
data class RendererAudioFormat(
  val bitsPerSample: Int,
  val sampleRate: Int,
  val numberOfChannels: Int,
  val commonFormat: String = "int16"
) {
  companion object {
    fun fromMap(formatMap: Map<String, Any?>): RendererAudioFormat? {
      val bitsPerSample = formatMap["bitsPerSample"] as? Int ?: 16
      val sampleRate = formatMap["sampleRate"] as? Int ?: 48000
      val numberOfChannels = formatMap["channels"] as? Int ?: 1
      val commonFormat = formatMap["commonFormat"] as? String ?: "int16"

      return RendererAudioFormat(bitsPerSample, sampleRate, numberOfChannels, commonFormat)
    }
  }
}
